### ISSUE TITLE
[Linux/Wine] Build 374ba5e is again broken on Wine (staging)

### ISSUE THREAD

**Issue by lorn10** [Cxbx-Reloaded/Cxbx-Reloaded#2347]
## Quick summary

Unfortunately it looks that build 374ba5e is again fatally broken on Wine. I have tested this on Wine 7.6. The previous CXBX-R version 696d498 and all others are working fine.

## Details

Out of my view this looks like a problem in the **building process** of CXBX-R because I have at the command line the following information:
```
wine: Call from 7B011586 to unimplemented function msvcp140_atomic_wait.dll.__std_atomic_wait_direct, aborting
wine: Call from 7B011586 to unimplemented function msvcp140_atomic_wait.dll.__std_atomic_notify_one_direct, aborting
```
These lines are **not** present at the older CXBX-R builds.

Note, this problem occurs regardless of whether native MS Visual C++ 2019 is installed through `winetricks -q vcrun2019`. As already mentioned in #2161 that option should be avoided anyway. MS Visual C++ 2019 is in newer Wine versions no longer fully installable (because of some compatibility problems) ergo some native Visual C++ 2019 parts are skipped during the installation with winetricks. However, that's not necessary anyway, CXBX-R worked until 374ba5e really well also with the built-in Visual C++ implementation of Wine.

### System Configuration:

```
test@test-iMac:~$ inxi -b
System:    Host: test-iMac Kernel: 5.13.0-39-generic x86_64 bits: 64 Desktop: KDE Plasma 5.18.8 
           Distro: Ubuntu 20.04.4 LTS (Focal Fossa) 
Machine:   Type: Desktop System: Apple product: iMac12,2 v: 1.0 serial: <superuser/root required> 
           Mobo: Apple model: Mac-942B59F58194171B v: iMac12,2 serial: <superuser/root required> UEFI: Apple 
           v: IM121.88Z.0047.B1F.1201241648 date: 01/24/12 
CPU:       Quad Core: Intel Core i5-2500S type: MCP speed: 1597 MHz min/max: 1600/3700 MHz 
Graphics:  Device-1: Intel 2nd Generation Core Processor Family Integrated Graphics driver: i915 v: kernel 
           Device-2: Advanced Micro Devices [AMD/ATI] Whistler [Radeon HD 6730M/6770M/7690M XT] driver: radeon v: kernel 
           Display: x11 server: X.Org 1.20.13 driver: modesetting,radeon FAILED: ati unloaded: fbdev,vesa 
           resolution: 2560x1440~60Hz 
           OpenGL: renderer: AMD TURKS (DRM 2.50.0 / 5.13.0-39-generic LLVM 14.0.0) 
           v: 3.3 Mesa 22.1.0-devel (git-c3ad133 2022-04-13 focal-oibaf-ppa) 
Network:   Device-1: Broadcom and subsidiaries NetXtreme BCM57765 Gigabit Ethernet PCIe driver: tg3 
           Device-2: Qualcomm Atheros AR93xx Wireless Network Adapter driver: ath9k 
Drives:    Local Storage: total: 961.01 GiB used: 114.70 GiB (11.9%) 
Info:      Processes: 210 Uptime: 2h 23m Memory: 7.74 GiB used: 1.60 GiB (20.6%) Shell: bash inxi: 3.0.38 
test@test-iMac:~$ 
```

### Additional Information (if any):

I noticed a similar problem with the PSX emulator **PCSX-Redux**. For a certain time it helped to copy the `msvcp140_atomic_wait.dll` file direly into the emulator folder. However, this worked just for some time and then that workaround stopped working. It looks that the `msvcp140_atomic_wait.dll` is no longer accepted by the emulator or by Wine. :wink:

**Comment by RadWolfie** [comment]
Your ticket originally closed because of " (if any)" was omitted out from "Additional Information (if any):" header requirement.
I made a correction and reopening this ticket.

**Comment by RadWolfie** [comment]
Since it is not a bug, but a missing implemention from Wine's end.

One thing I'll like to add according to details section' log, those two functions are used by semaphore's acquire and release functions.

**Comment by lorn10** [comment]
Okay, so this is then only a problem of newer Wine versions **beyond 7.0**. I also have installed vcrun2019 (and d3dcompiler_47) but it still crashes with an unhandled exception:
![Screenshot_cxbxr_374ba5e_unhandled_exception](https://user-images.githubusercontent.com/9931771/163343837-bafa8a1b-037d-41ab-bb5e-6c5a28afb3c2.png)

**Comment by RadWolfie** [comment]
> Okay, so this is then only a problem of newer Wine versions **beyond 7.0**. I also have installed vcrun2019 (and d3dcompiler_47) but it still crashes with an unhandled exception: [Screenshot_cxbxr_374ba5e_unhandled_exception](https://user-images.githubusercontent.com/9931771/163343837-bafa8a1b-037d-41ab-bb5e-6c5a28afb3c2.png)

Issue you mentioned appear to be different issue. I had never seen `0x80000100` error code before and it's not part of NTSTATUS's error codes. I had a looked into Wine's source code and found it is [`EXCEPTION_WINE_STUB`](https://source.winehq.org/git/wine.git/blob/07f22e20d73ce0a3c758d331400e765795d675a0:/include/winternl.h#l3491).

**Comment by lorn10** [comment]
But `0x7B011586` seems related to the `Call from 7B011586` message at the CLI, right?

**Comment by RadWolfie** [comment]
Correct, you will need to check which module is loaded in the address.

**Comment by RadWolfie** [comment]
> **Addition**: Maybe someone can build a VS2019 or VS2022 CXBX-R package without that new semaphore C++ functionality?

It no longer matter if it's VS2019 or VS2022, however... VS2019 doesn't have bugfix(es) for C++20 which VS2022 has at the moment. Also, we are going forward with C++20 standard. It is up to Wine community to implement them. Beside, semaphore doesn't pose any problem with stable wine release build as of 7.0 with vcrun2019 installed.

**Comment by lorn10** [comment]
Here is the backtrace of this error message. I was able to provoke it when I chose the "abort" (Abbrechen) button: :wink: 

[cxbxr_374ba5e_unhandled_exception_backtrace.txt](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/files/8488132/cxbxr_374ba5e_unhandled_exception_backtrace.txt)

Yes, with Wine 7.0 `vcrun2019` seems to work. Exist there any chance that I will get this work also with `msvcp140_atomic_wait.dll` manually placed into the CXBX-R folder? As mentioned, this helped for some time with PCSX-Redux. And maybe my `msvcp140_atomic_wait.dll` became outdated, it has version 14.29.30037.0.

**Comment by RadWolfie** [comment]
I used PlayOnLinux software with Wine installed directly on linux machine. vcrun2019 installer is `14.29.30135.0` which is newer build than your's. I can confirm `msvcp140_atomic_wait.dll` is properly installed under Wine 7.0's virtual drive directory. So, you do not need to manually place `msvcp140_atomic_wait.dll`.

This type of problem should be report to wine. And we do not support any Wine staging builds. We're incline to close this ticket unless the problem pursuit under **stable** 7.0 build.

---

With that said, #2161 is mainly for to update our readme file to remove vcrun2017 requirement. Now, vcrun2019 is a requirement in order to run titles.

**Comment by PatrickvL** [comment]
Just a thought : this seems to have started with the introduction of `std::binary_semaphore` type, so it could be considered to not use this type and instead use the previous proposal (checking a boolean in a while loop).

**Comment by lorn10** [comment]
Okay, will close this issue for now. Perhaps this will be relevant again with **Wine 8.0**. :wink:

**Comment by lorn10** [comment]
For the record, there exist now a bug report at Wine about this missing C++20 Semaphore matter, see [Bug 52821](https://bugs.winehq.org/show_bug.cgi?id=52821).

**Comment by lorn10** [comment]
Great, I figured it out how this new "C++20 Semaphore" CXBX-R build 374ba5e works with Wine devel 7.6. :+1: 

A short HowTo:

 - rename Wines `msvcp140_atomic_wait.dll` in `System32` and `SysWOW64` to `msvcp140_atomic_waitWINE.dll` (or else)
 - copy the original MS `msvcp140_atomic_wait.dll` into the corresponding folder, the 64bit into `System32` and the 32bit into `SysWOW64`
 - start up `winecfg` from the CLI and select "Libraries", add here a new one called `msvcp140_atomic_wait` (without the dll extension)

There should be now a new entry "msvcp140_atomic_wait (Native, Builtin)". Exit winecfg and start up CXBX-R build 374ba5e. It should work again. I have tested this also with PCSX-Redux nightly 8457.

I grabbed the latest msvcp140_atomic_wait.dll file version from my last Windows 7 computer where I have installed the current MSVC runtime library from [here](https://docs.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170). It gave me file version `14.31.31103.0`.

And @RadWolfie, `vcrun2019` is definitely **not** a requirement to run newer CXBX-R builds. All more recent builds until 696d498 worked almost perfectly well for me **without it** (since around Wine 6.7). This changed with 374ba5e which requires now the "msvcp140_atomic_wait.dll" tweak regardless if `vcrun2019` is already present or not. :wink:

**Comment by RadWolfie** [comment]
From WineHQ's ticket:
> I think your modification in winetricks will for sure work. I would test that if I could but I am currently not so familiar in building stuff like this at my own.

@lorn10, winetricks is a batch script. All you need to do is download winetricks file, link provided by Austin, and mark as binary file in order to execute it. No building requirement involve.

**Comment by lorn10** [comment]
I have tested the `msvcp140_atomic_wait.dll` winetricks changes provided by @austin987 and indeed it worked as assumed by me. :+1: 

![winetricks_msvcp140_atomic_wait_test](https://user-images.githubusercontent.com/9931771/164023991-97a222c1-8dce-4187-a3cf-9a606a7ce2b2.png)

However, in the end I had to modify those two lines 12934 & 12962 in my own winetricks script. So yes, I "coded" a little bit.:wink:

**Comment by lorn10** [comment]
A final great update to this topic. The problem is resolved in **Wine 7.8 devel** (and most likely also Wine 7.8 staging). :+1: 

The reason is that the needed "msvcp140_atomic_wait.dll.__std_atomic_wait_direct" function was successfully added to the built-in "msvcp140_atomic_wait.dll" file. Therefore the native "msvcp140_atomic_wait.dll" file is no longer needed. CC @RadWolfie 

I can confirm this for CXBX-Reloaded build 374ba5e which uses C++20 Semaphores. It is now working also with the built-in variant fine, no difference to the native original MS file.

So this change will now make the winetricks `vcrun2019` option again fully optional.
